File: internal/flightident/flightident_test.go
package flightident

import "testing"

func TestNormalise(t *testing.T) {
	cases := map[string]string{
		"BA286":    "BA286",
		"ba 286":   "BA286",
		" BA 286 ": "BA286",
		"BA  286":  "BA286",
		"lh\t441":  "LH441",
		"":         "",
	}
	for in, want := range cases {
		if got := Normalise(in); got != want {
			t.Errorf("Normalise(%q) = %q, want %q", in, got, want)
		}
	}
}

func TestSplit(t *testing.T) {
	cases := []struct {
		in                      string
		airline, number, suffix string
		ok                      bool
	}{
		// The ordinary two-letter designator, spaced or not.
		{"BA286", "BA", "286", "", true},
		{"BA 286", "BA", "286", "", true},
		{"ba286", "BA", "286", "", true},

		// Designators carrying a digit: the whole point of issue #118. A
		// letters-first rule would read these as airline "U" flight 21234.
		{"U21234", "U2", "1234", "", true},
		{"U2 1234", "U2", "1234", "", true},
		{"U287", "U2", "87", "", true},
		{"9W420", "9W", "420", "", true},
		{"4U2678", "4U", "2678", "", true},

		// Leading zeros are cosmetic; the operational suffix is not.
		{"BA0087", "BA", "87", "", true},
		{"BA00087", "BA", "87", "", true},
		{"BA286A", "BA", "286", "A", true},

		// Three-letter ICAO designators, only when two characters won't do.
		{"BAW123", "BAW", "123", "", true},
		{"DLH1", "DLH", "1", "", true},

		// Not flight numbers.
		{"", "", "", "", false},
		{"BA", "", "", "", false},
		{"12345", "", "", "", false},
		{"1234", "", "", "", false},
		{"BA0000", "", "", "", false},
		{"AC12345", "", "", "", false},
		{"GIBBERISH", "", "", "", false},
		{"BA-286", "", "", "", false},
	}
	for _, c := range cases {
		t.Run(c.in, func(t *testing.T) {
			airline, number, suffix, ok := Split(c.in)
			if ok != c.ok || airline != c.airline || number != c.number || suffix != c.suffix {
				t.Errorf("Split(%q) = (%q, %q, %q, %v), want (%q, %q, %q, %v)",
					c.in, airline, number, suffix, ok, c.airline, c.number, c.suffix, c.ok)
			}
		})
	}
}

func TestValid(t *testing.T) {
	if !Valid("lh 441") {
		t.Error("a spaced, lower-case flight number should be valid")
	}
	if Valid("nonsense") {
		t.Error("a word is not a flight number")
	}
}

func TestCanonical(t *testing.T) {
	cases := []struct {
		in, want string
		ok       bool
	}{
		{"BA87", "BA0087", true},
		{"BA 87", "BA0087", true},
		{"BA0087", "BA0087", true},
		{"U287", "U20087", true}, // not "U0287": the designator owns that digit
		{"9W420", "9W0420", true},
		{"BA286A", "BA0286A", true},
		{"GIBBERISH", "", false},
	}
	for _, c := range cases {
		got, ok := Canonical(c.in)
		if ok != c.ok || got != c.want {
			t.Errorf("Canonical(%q) = (%q, %v), want (%q, %v)", c.in, got, ok, c.want, c.ok)
		}
	}
}

File: web/src/lib/flight-ident.test.ts
import { describe, expect, it } from 'vitest';

import { airlineOf, normaliseIdent, splitIdent } from './flight-ident';

describe('normaliseIdent', () => {
  it('upper-cases and strips whitespace', () => {
    expect(normaliseIdent('ba 286')).toBe('BA286');
    expect(normaliseIdent('  BA  286 ')).toBe('BA286');
    expect(normaliseIdent('')).toBe('');
  });
});

describe('splitIdent', () => {
  it('splits an ordinary two-letter designator, spaced or not', () => {
    expect(splitIdent('BA286')).toEqual({ airline: 'BA', number: '286', suffix: '' });
    expect(splitIdent('ba 286')).toEqual({ airline: 'BA', number: '286', suffix: '' });
  });

  // Issue #118: a letters-only prefix rule reads "U21234" as airline "U".
  it('keeps a digit that belongs to the designator', () => {
    expect(splitIdent('U21234')).toEqual({ airline: 'U2', number: '1234', suffix: '' });
    expect(splitIdent('9W420')).toEqual({ airline: '9W', number: '420', suffix: '' });
    expect(splitIdent('4U2678')).toEqual({ airline: '4U', number: '2678', suffix: '' });
  });

  it('strips leading zeros and separates the operational suffix', () => {
    expect(splitIdent('BA0087')).toEqual({ airline: 'BA', number: '87', suffix: '' });
    expect(splitIdent('BA286A')).toEqual({ airline: 'BA', number: '286', suffix: 'A' });
  });

  it('falls back to a three-letter ICAO designator', () => {
    expect(splitIdent('BAW123')).toEqual({ airline: 'BAW', number: '123', suffix: '' });
  });

  it('rejects what is not a flight number', () => {
    for (const s of ['', 'BA', '12345', '1234', 'BA0000', 'AC12345', 'GIBBERISH', 'BA-286']) {
      expect(splitIdent(s), s).toBeNull();
    }
  });
});

describe('airlineOf', () => {
  it('returns the designator, or null', () => {
    expect(airlineOf('U2 8021')).toBe('U2');
    expect(airlineOf('nonsense')).toBeNull();
  });
});
